File: android/app/src/main/java/fr/xebia/magritte/classifier/ClassifierPresenter.java
package fr.xebia.magritte.classifier;

import android.content.Context;
import android.content.res.Configuration;
import android.graphics.Bitmap;
import android.util.Log;

import java.util.List;
import java.util.Locale;
import java.util.concurrent.Callable;

import fr.xebia.magritte.model.FruitResource;
import io.reactivex.Single;
import io.reactivex.SingleObserver;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

public class ClassifierPresenter implements ClassifierContract.Presenter {

    private static final double MATCH_THRESHOLD = 0.9;
    private static final String TAG = ClassifierPresenter.class.getSimpleName();

    private Locale desiredLocale;

    private ClassifierContract.View view;
    private Classifier classifier;
    private Classifier.Recognition topMatch;

    public ClassifierPresenter(ClassifierContract.View view, Locale desiredLocale, Classifier classifier) {
        this.view = view;
        this.classifier = classifier;
        this.desiredLocale = desiredLocale;
    }

    @Override
    public void recognizeImage(final Bitmap bitmap) {
        Single<List<Classifier.Recognition>> recognitionSingle = Single.fromCallable(new Callable<List<Classifier.Recognition>>() {

            @Override
            public List<Classifier.Recognition> call() throws Exception {
                return classifier.recognizeImage(bitmap);
            }
        });

        recognitionSingle
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new SingleObserver<List<Classifier.Recognition>>() {

                    @Override
                    public void onSubscribe(Disposable d) {

                    }

                    @Override
                    public void onSuccess(List<Classifier.Recognition> results) {
                        view.displayRecognitions(results);

                        Classifier.Recognition topMatch = getTopMatch(results);
                        if (topMatch != null) {
                            proceedTopMatch(topMatch);
                        }
                    }

                    @Override
                    public void onError(Throwable e) {

                    }
                });
    }

    private FruitResource getFruitResult(Classifier.Recognition recognition) {
        try {
            return FruitResource.valueOf(recognition.getTitle().toUpperCase());
        } catch (IllegalArgumentException e) {
            Log.d(TAG, "No matching");
            return null;
        }
    }

    private void proceedTopMatch(Classifier.Recognition topMatch) {
        if (this.topMatch == null || !this.topMatch.getTitle().equals(topMatch.getTitle())) {
            FruitResource fruit = getFruitResult(topMatch);
            if (fruit != null) {
                view.displayTopMatch(fruit);
                view.speakResult(getLocalizedString(desiredLocale, fruit.getTitleRes()));
            }
            this.topMatch = topMatch;
        }
    }

    public String getLocalizedString(Locale desiredLocale, int resId) {
        Context context = view.getContext();
        Configuration conf = context.getResources().getConfiguration();
        conf = new Configuration(conf);
        conf.setLocale(desiredLocale);
        Context localizedContext = context.createConfigurationContext(conf);
        return localizedContext.getResources().getString(resId);
    }

    private Classifier.Recognition getTopMatch(List<Classifier.Recognition> recognitions) {
        for (Classifier.Recognition recognition : recognitions) {
            if (recognition.getConfidence() > MATCH_THRESHOLD) {
                return recognition;
            }
        }
        return null;
    }
}

File: android/app/src/main/java/fr/xebia/magritte/classifier/ClassifierPresenter.kt
package fr.xebia.magritte.classifier

import android.content.res.Configuration
import android.graphics.Bitmap
import android.util.Log
import fr.xebia.magritte.model.FruitResource
import io.reactivex.Single
import io.reactivex.SingleObserver
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import java.util.*

class ClassifierPresenter(private val view: ClassifierContract.View, private val desiredLocale: Locale, private val classifier: Classifier) : ClassifierContract.Presenter {
    private var topMatch: Classifier.Recognition? = null

    override fun recognizeImage(bitmap: Bitmap) {
        val recognitionSingle = Single.fromCallable { classifier.recognizeImage(bitmap) }

        recognitionSingle
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : SingleObserver<MutableList<Classifier.Recognition>> {

                    override fun onSubscribe(d: Disposable) {}
                    override fun onSuccess(results: MutableList<Classifier.Recognition>) {
                        view.displayRecognitions(results)

                        val topMatch = getTopMatch(results)
                        if (topMatch != null) {
                            proceedTopMatch(topMatch)
                        }
                    }

                    override fun onError(e: Throwable) {}
                })
    }

    private fun getFruitResult(recognition: Classifier.Recognition): FruitResource {
        return try {
            FruitResource.valueOf(recognition.title.toUpperCase())
        } catch (e: IllegalArgumentException) {
            Log.d(TAG, "No matching")
            FruitResource.UNKNOWN
        }
    }

    private fun proceedTopMatch(topMatch: Classifier.Recognition?) {
        if (this.topMatch == null || this.topMatch!!.title != topMatch!!.title) {
            val fruit = topMatch?.let { getFruitResult(it) }
            if (fruit != FruitResource.UNKNOWN) {
                fruit?.let {
                    view.displayTopMatch(it)
                    view.speakResult(getLocalizedString(desiredLocale, fruit.titleRes))
                }
            }
            this.topMatch = topMatch
        }
    }

    override fun getLocalizedString(locale: Locale, resId: Int): String {
        val context = view.context
        var conf = context.resources.configuration
        conf = Configuration(conf)
        conf.setLocale(locale)
        val localizedContext = context.createConfigurationContext(conf)
        return localizedContext.resources.getString(resId)
    }

    private fun getTopMatch(recognitions: List<Classifier.Recognition>): Classifier.Recognition? {
        return recognitions.firstOrNull { it.confidence > MATCH_THRESHOLD }
    }

    companion object {

        private val MATCH_THRESHOLD = 0.85
        private val TAG = ClassifierPresenter::class.java.simpleName
    }
}
